src/java.base/share/classes/jdk/internal/foreign/abi/BufferStack.java

```diff
@@ -0,0 +1,122 @@
+ * questions.
+ */
+package jdk.internal.foreign.abi;
+
+import jdk.internal.foreign.SlicingAllocator;
+import jdk.internal.misc.CarrierThreadLocal;
+import jdk.internal.vm.annotation.ForceInline;
+
+import java.lang.foreign.Arena;
+import java.lang.foreign.MemorySegment;
+import java.lang.foreign.SegmentAllocator;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class BufferStack {
+    private final long size;
+
+    public BufferStack(long size) {
+        this.size = size;
+    }
+
+    private final ThreadLocal<PerThread> tl = new CarrierThreadLocal<>() {
+        @Override
+        protected PerThread initialValue() {
+            return new PerThread(size);
+        }
+    };
+
+    @ForceInline
+    public Arena pushFrame(long size, long byteAlignment) {
+        return tl.get().pushFrame(size, byteAlignment);
+    }
+
+    private static final class PerThread {
+        private final ReentrantLock lock = new ReentrantLock();
+        private final SlicingAllocator stack;
+
+        public PerThread(long size) {
+            this.stack = new SlicingAllocator(Arena.ofAuto().allocate(size));
+        }
+
+        @ForceInline
+        public Arena pushFrame(long size, long byteAlignment) {
+            boolean needsLock = Thread.currentThread().isVirtual() && !lock.isHeldByCurrentThread();
+            if (needsLock && !lock.tryLock()) {
+                // Rare: another virtual thread on the same carrier competed for acquisition.
+                return Arena.ofConfined();
+            }
+            if (!stack.canAllocate(size, byteAlignment)) {
+                if (needsLock) lock.unlock();
+                return Arena.ofConfined();
+            }
+
+            return new Frame(needsLock, size, byteAlignment);
+        }
+
+        private class Frame implements Arena {
+            private final boolean locked;
+            private final long parentOffset;
+            private final long topOfStack;
+            private final Arena scope = Arena.ofConfined();
+            private final SegmentAllocator frame;
+
+            @SuppressWarnings("restricted")
+            public Frame(boolean locked, long byteSize, long byteAlignment) {
+                this.locked = locked;
+
+                parentOffset = stack.currentOffset();
+                MemorySegment frameSegment = stack.allocate(byteSize, byteAlignment);
+                topOfStack = stack.currentOffset();
+                frame = new SlicingAllocator(frameSegment.reinterpret(scope, null));
+            }
+
+            private void assertOrder() {
+                if (topOfStack != stack.currentOffset())
+                    throw new IllegalStateException("Out of order access: frame not top-of-stack");
+            }
+
+            @Override
+            @SuppressWarnings("restricted")
+            public MemorySegment allocate(long byteSize, long byteAlignment) {
+                return frame.allocate(byteSize, byteAlignment);
```

Comment:
Should this also check order?

Reply:
We could, in the sense that an allocation in a lower stack frame seems suspicious, but technically it is completely legal. The frame has been allocated and is sliced to the requested size, and is guaranteed as long as the Frame's arena hasn't been closed, no matter whether other frames are on top:
```
frame1 = pushFrame(256);
frame2 = pushFrame(256);
<<frame1 can safely allocate up to 256>>
```

Reply:
> We could, in the sense that an allocation in a lower stack frame seems suspicious, but technically it is completely legal. The frame has been allocated and is sliced to the requested size, and is guaranteed as long as the Frame's arena hasn't been closed, no matter whether other frames are on top:
> 
> ```
> frame1 = pushFrame(256);
> frame2 = pushFrame(256);
> <<frame1 can safely allocate up to 256>>
> ```

Ok, in this particular design it's ok because you allocate up front. So each frame allocate in its own space.